### PR TITLE
Make JitDump label output multi-thread safe

### LINKED ISSUE #10984: Fix JitDump label output to be multi-thread safe
We emit labels using a global static `Compiler::s_compMethodsCount`, e.g.:
```
printf("G_M%03u_IG%02u", Compiler::s_compMethodsCount, id->idAddr()->iiaIGlabel->igNum);
```

That’s dangerous if two JITs are running at the same time, and generating JitDumps. We should change to using a Compiler member, and then change:
```
#ifdef DEBUG
    /* Give the function a unique number */

    if (opts.disAsm || opts.dspEmit || verbose)
    {
        s_compMethodsCount = ~info.compMethodHash() & 0xffff;
    }
    else
    {
        s_compMethodsCount++;
    }
#endif
```
To:
```
#ifdef DEBUG
    /* Give the function a unique number */

    if (opts.disAsm || opts.dspEmit || verbose)
    {
        compMethodNumber = ~info.compMethodHash() & 0xffff;
    }
    else
    {
        compMethodNumber = s_compMethodsCount;
        s_compMethodsCount++;
    }
#endif
```
And the calls to:
```
printf("G_M%03u_IG%02u", compMethodNumber, id->idAddr()->iiaIGlabel->igNum);
```

Some places we access `s_compMethodsCount` currently might not have a `Compiler` pointer, so it might be a little more work than this.

category:implementation
theme:debug-dumps
skill-level:beginner
cost:small

### PR DESCRIPTION
**Pull request by kunalspathak**
- Interlock the static variable when incrementing
- Fixed 2 broken links inside botr's `ryujit-overview.md` doc.

Fixes: #10984

### CODE REVIEW

**Comment by kunalspathak** [comment]
cc @BruceForstall

**Comment by kunalspathak** [comment]
Looks like `<winnt.h>` doesn't get included. I will take a look tomorrow.

**Comment by BruceForstall** [comment]
The issue isn't really the lack of Interlocked; it's that lots of places use `s_compMethodsCount`, but that could be updated by concurrent JIT invocations. We need to use an instance variable -- probably on our single-instance "Compiler" class, but potentially also a TLS variable.

As for winnt.h, we don't use that because we must be portable to Linux/Mac. We wrap everything with either utilcode.h/winwrap.h, or for Linux we use src\coreclr\src\pal\inc\pal.h.

File: src/coreclr/src/jit/compiler.h [hunk 1]
```diff
@@ -8979,10 +8979,16 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 //-------------------------- Global Compiler Data ------------------------------------
 
 #ifdef DEBUG
-    static unsigned s_compMethodsCount; // to produce unique label names
-    unsigned        compGenTreeID;
-    unsigned        compStatementID;
-    unsigned        compBasicBlockID;
+private:
+    static LONG s_compMethodID; // to produce unique label names
```
Comment (BruceForstall) [inline src/coreclr/src/jit/compiler.h]: Shouldn't this still be `s_compMethodsCount`? After all, it is still a static global count of the number of methods the JIT has compiled. We just happen to often initialize the "method ID" using this.

**Review by BruceForstall — COMMENTED** [review]
One question

**Comment by BruceForstall** [comment]
cc @dotnet/jit-contrib

File: src/coreclr/src/jit/compiler.h (hunk 1 above)
Comment (kunalspathak) [reply src/coreclr/src/jit/compiler.h]: You are right, it is still a static global method count that are JITted but it was (and is) used in producing a unique label names. In fact, I named the instance variable as `compMethodID` and it felt little odd to have name inconsistency during assignment. But I don't have any preference one way or other. I will go ahead and revert the name of this variable.

Comment (kunalspathak) [reply src/coreclr/src/jit/compiler.h]: done.

**Review by erozenfeld — APPROVED** [review]
LGTM

**Review by BruceForstall — APPROVED** [review]

File: src/coreclr/src/jit/compiler.cpp [hunk 2]
```diff
@@ -1299,7 +1299,7 @@ size_t genFlowNodeCnt;
 
 #ifdef DEBUG
 /* static */
-unsigned Compiler::s_compMethodsCount = 0; // to produce unique label names
+LONG Compiler::s_compMethodsCount = 0; // to produce unique label names
```
Comment (sandreenko) [inline src/coreclr/src/jit/compiler.cpp]: have not seen many `LONG` in the jit, I am not sure in which case we use them.

**Review by sandreenko — APPROVED** [review]
LGTM

File: src/coreclr/src/jit/compiler.cpp (hunk 2 above)
Comment (BruceForstall) [reply src/coreclr/src/jit/compiler.cpp]: @kunalspathak Presumably it chosen because it matches the signature of InterlockedIncrement

Comment (kunalspathak) [reply src/coreclr/src/jit/compiler.cpp]: That's correct.

**Comment by benaadams** [comment]
I think there is a missed method (perhaps from overlapping merges)

https://github.com/dotnet/runtime/blob/4f84429fb44801b841c052505935a3cd14da18e2/src/coreclr/src/jit/codegencommon.cpp#L11307-L11312

Not sure runtime currently compiles?

**Pull request merged**